lib/plug/adapters/cowboy2.ex

```diff
@@ -0,0 +1,306 @@
+      Defaults to false.
+
+    * `:timeout` - Time in ms with no requests before Cowboy closes the connection.
+      Defaults to 5000ms.
+
+    * `:protocol_options` - Specifies remaining protocol options,
+      see [Cowboy2 protocol docs](http://ninenines.eu/docs/en/cowboy/2.0/manual/cowboy_protocol/).
+
+  All other options are given to the underlying transport.
+  """
+
+  require Logger
+
+  # Made public with @doc false for testing.
+  @doc false
+  def args(scheme, plug, opts, cowboy_options) do
+    {cowboy_options, non_keyword_options} =
+      Enum.partition(cowboy_options, &is_tuple(&1) and tuple_size(&1) == 2)
+
+    cowboy_options
+    |> Keyword.put_new(:max_connections, 16_384)
+    |> Keyword.put_new(:ref, build_ref(plug, scheme))
+    |> Keyword.put_new(:dispatch, cowboy_options[:dispatch] || dispatch_for(plug, opts))
+    |> set_compress()
+    |> normalize_cowboy_options(scheme)
+    |> to_args(non_keyword_options)
+  end
+
+  @doc """
+  Runs cowboy under http.
+
+  ## Example
+
+      # Starts a new interface
+      Plug.Adapters.Cowboy2.http MyPlug, [], port: 80
+
+      # The interface above can be shutdown with
+      Plug.Adapters.Cowboy2.shutdown MyPlug.HTTP
+
+  """
+  @spec http(module(), Keyword.t, Keyword.t) ::
+        {:ok, pid} | {:error, :eaddrinuse} | {:error, term}
+  def http(plug, opts, cowboy_options \\ []) do
+    run(:http, plug, opts, cowboy_options)
+  end
+
+  @doc """
+  Runs cowboy under https.
+
+  Besides the options described in the module documentation,
+  this module also accepts all options defined in [the `ssl`
+  erlang module] (http://www.erlang.org/doc/man/ssl.html),
+  like keyfile, certfile, cacertfile, dhfile and others.
+
+  The certificate files can be given as a relative path.
+  For such, the `:otp_app` option must also be given and
+  certificates will be looked from the priv directory of
+  the given application.
+
+  ## Example
+
+      # Starts a new interface
+      Plug.Adapters.Cowboy2.https MyPlug, [],
+        port: 443,
+        password: "SECRET",
+        otp_app: :my_app,
+        keyfile: "priv/ssl/key.pem",
+        certfile: "priv/ssl/cert.pem",
+        dhfile: "priv/ssl/dhparam.pem"
+
+      # The interface above can be shutdown with
+      Plug.Adapters.Cowboy2.shutdown MyPlug.HTTPS
+
+  """
+  @spec https(module(), Keyword.t, Keyword.t) ::
+        {:ok, pid} | {:error, :eaddrinuse} | {:error, term}
+  def https(plug, opts, cowboy_options \\ []) do
+    Application.ensure_all_started(:ssl)
+    run(:https, plug, opts, cowboy_options)
+  end
+
+  @doc """
+  Shutdowns the given reference.
+  """
+  def shutdown(ref) do
+    :cowboy.stop_listener(ref)
+  end
+
+  @doc """
+  Returns a child spec to be supervised by your application.
+
+  This function returns the old child specs used by early OTP
+  and Elixir versions. See `child_spec/1` for the Elixir v1.5
+  based child specifications.
+  """
+  def child_spec(scheme, plug, opts, cowboy_options \\ []) do
```

Comment:
I am wondering if we can remove this API and focus on `child_spec/1` exclusively.

Reply:
We certainly can. The only down side is people not being able to simply replace `Plug.Adapters.Cowboy.child_spec` with `Plug.Adapters.Cowboy2.child_spec`. However since we will include a cowboy 2 upgrade guide with the next release, I don't think that's a problem. Even if you are using Elixir 1.4 you can still use child_spec/1.